From 7db5a1c7ed61378c06841efc3d4da8ca98ce2eba Mon Sep 17 00:00:00 2001
From: Tom Hunze <dev@thunze.de>
Date: Sat, 30 Aug 2025 22:37:08 +0200
Subject: [PATCH 1/4] python313Packages.i3ipc: fix build

Upstream hasn't been very active in the past years.

1. pytest 8.4 introduced a breaking change where async tests fail
    instead of generating a warning when there is no async-aware plugin
    to handle them.
  - Changelog: https://github.com/pytest-dev/pytest/releases/tag/8.4.0
  - Fixed by adding pytest-asyncio as a dependency.

2. Upstream expects a very old version of pytest-asyncio. To make the
    tests work with latest pytest-asyncio I added a patch which:
  - Correctly decorates the only async fixture `i3` with
      `@pytest_asyncio.fixture`.
  - Configures `loop_scope='class'` for all async tests and the async
      fixture `i3` to match the existing `scope='class'` on `i3`.
---
 .../python-modules/i3ipc/default.nix          | 10 +++++++
 .../i3ipc/fix-async-tests.patch               | 28 +++++++++++++++++++
 2 files changed, 38 insertions(+)
 create mode 100644 pkgs/development/python-modules/i3ipc/fix-async-tests.patch

diff --git a/pkgs/development/python-modules/i3ipc/default.nix b/pkgs/development/python-modules/i3ipc/default.nix
index febac35bc4a452e213c63b62cbeec038d1093e03..e63dc1f68c55a120ffd13ab6c8db110d83fc0d96 100644
--- a/pkgs/development/python-modules/i3ipc/default.nix
+++ b/pkgs/development/python-modules/i3ipc/default.nix
@@ -4,6 +4,7 @@
   fetchFromGitHub,
   xorg,
   pytest,
+  pytest-asyncio,
   pytest-xvfb,
   i3,
   xlib,
@@ -23,12 +24,21 @@ buildPythonPackage rec {
     rev = "v${version}";
     sha256 = "13bzs9dcv27czpnnbgz7a037lm8h991c8gk0qzzk5mq5yak24715";
   };
+
+  patches = [
+    # Upstream expects a very old version of pytest-asyncio. This patch correctly
+    # decorates async fixtures using pytest-asyncio and configures `loop_scope`
+    # where needed.
+    ./fix-async-tests.patch
+  ];
+
   propagatedBuildInputs = [ xlib ];
 
   fontsConf = makeFontsConf { fontDirectories = [ ]; };
   FONTCONFIG_FILE = fontsConf; # Fontconfig error: Cannot load default config file
   nativeCheckInputs = [
     pytest
+    pytest-asyncio
     xdpyinfo
     pytest-xvfb
     xorg.xvfb
diff --git a/pkgs/development/python-modules/i3ipc/fix-async-tests.patch b/pkgs/development/python-modules/i3ipc/fix-async-tests.patch
new file mode 100644
index 0000000000000000000000000000000000000000..41407c1d4878f46bad91944d70d82e75faac4f5c
--- /dev/null
+++ b/pkgs/development/python-modules/i3ipc/fix-async-tests.patch
@@ -0,0 +1,28 @@
+diff --git a/test/aio/ipctest.py b/test/aio/ipctest.py
+index 88e4cda..3d0fd9c 100644
+--- a/test/aio/ipctest.py
++++ b/test/aio/ipctest.py
+@@ -1,5 +1,6 @@
+ from subprocess import Popen
+ import pytest
++import pytest_asyncio
+ 
+ from i3ipc.aio import Connection
+ from i3ipc import CommandReply
+@@ -19,7 +20,7 @@ class IpcTest:
+     def event_loop(self):
+         return asyncio.get_event_loop()
+ 
+-    @pytest.fixture(scope='class')
++    @pytest_asyncio.fixture(scope='class', loop_scope='class')
+     async def i3(self):
+         process = Popen(['i3', '-c', 'test/i3.config'])
+         # wait for i3 to start up
+diff --git a/pytest.ini b/pytest.ini
+index 1ea6b80..788bdac 100644
+--- a/pytest.ini
++++ b/pytest.ini
+@@ -1,2 +1,3 @@
+ [pytest]
+ timeout = 5
++asyncio_default_test_loop_scope = class

From 99a5b16b7606934b54ae0a2181332a790cfd7a37 Mon Sep 17 00:00:00 2001
From: Tom Hunze <dev@thunze.de>
Date: Sat, 30 Aug 2025 23:25:53 +0200
Subject: [PATCH 2/4] python313Packages.i3ipc: add pytest-timeout dependency

---
 pkgs/development/python-modules/i3ipc/default.nix | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/pkgs/development/python-modules/i3ipc/default.nix b/pkgs/development/python-modules/i3ipc/default.nix
index e63dc1f68c55a120ffd13ab6c8db110d83fc0d96..0b8411d7abe32e301a0dc06a874ab8f4b6e1aa26 100644
--- a/pkgs/development/python-modules/i3ipc/default.nix
+++ b/pkgs/development/python-modules/i3ipc/default.nix
@@ -5,6 +5,7 @@
   xorg,
   pytest,
   pytest-asyncio,
+  pytest-timeout,
   pytest-xvfb,
   i3,
   xlib,
@@ -40,6 +41,7 @@ buildPythonPackage rec {
     pytest
     pytest-asyncio
     xdpyinfo
+    pytest-timeout
     pytest-xvfb
     xorg.xvfb
     i3

From 3d53ad129b680202045ecc77bd36d273577268f3 Mon Sep 17 00:00:00 2001
From: Tom Hunze <dev@thunze.de>
Date: Sat, 30 Aug 2025 23:27:03 +0200
Subject: [PATCH 3/4] python313Packages.i3ipc: refactor

---
 .../python-modules/i3ipc/default.nix          | 62 ++++++++++---------
 1 file changed, 34 insertions(+), 28 deletions(-)

diff --git a/pkgs/development/python-modules/i3ipc/default.nix b/pkgs/development/python-modules/i3ipc/default.nix
index 0b8411d7abe32e301a0dc06a874ab8f4b6e1aa26..c3458c3bc98fe729d14d8be7cb06f15e5eb97c4e 100644
--- a/pkgs/development/python-modules/i3ipc/default.nix
+++ b/pkgs/development/python-modules/i3ipc/default.nix
@@ -2,28 +2,28 @@
   lib,
   buildPythonPackage,
   fetchFromGitHub,
-  xorg,
-  pytest,
+  coreutils,
+  setuptools,
+  xlib,
+  fontconfig,
+  pytestCheckHook,
   pytest-asyncio,
   pytest-timeout,
   pytest-xvfb,
   i3,
-  xlib,
-  xdpyinfo,
-  makeFontsConf,
-  coreutils,
+  xorg,
 }:
 
 buildPythonPackage rec {
   pname = "i3ipc";
   version = "2.2.1";
-  format = "setuptools";
+  pyproject = true;
 
   src = fetchFromGitHub {
-    owner = "acrisci";
+    owner = "altdesktop";
     repo = "i3ipc-python";
-    rev = "v${version}";
-    sha256 = "13bzs9dcv27czpnnbgz7a037lm8h991c8gk0qzzk5mq5yak24715";
+    tag = "v${version}";
+    hash = "sha256-JRwipvIF1zL/x2A+xEJKEFV6BlDnv2Xt/eyIzVrSf40=";
   };
 
   patches = [
@@ -33,34 +33,40 @@ buildPythonPackage rec {
     ./fix-async-tests.patch
   ];
 
-  propagatedBuildInputs = [ xlib ];
+  postPatch = ''
+    substituteInPlace test/i3.config \
+      --replace-fail /bin/true ${coreutils}/bin/true
+  '';
+
+  build-system = [ setuptools ];
+  dependencies = [ xlib ];
+
+  # Fontconfig error: Cannot load default config file
+  env.FONTCONFIG_FILE = "${fontconfig.out}/etc/fonts/fonts.conf";
 
-  fontsConf = makeFontsConf { fontDirectories = [ ]; };
-  FONTCONFIG_FILE = fontsConf; # Fontconfig error: Cannot load default config file
   nativeCheckInputs = [
-    pytest
+    pytestCheckHook
     pytest-asyncio
-    xdpyinfo
     pytest-timeout
     pytest-xvfb
-    xorg.xvfb
     i3
+    xorg.xdpyinfo
+    xorg.xvfb
   ];
 
-  postPatch = ''
-    substituteInPlace test/i3.config \
-      --replace /bin/true ${coreutils}/bin/true
-  '';
+  disabledTestPaths = [
+    # Timeout
+    "test/test_shutdown_event.py::TestShutdownEvent::test_shutdown_event_reconnect"
+    "test/aio/test_shutdown_event.py::TestShutdownEvent::test_shutdown_event_reconnect"
+  ];
 
-  checkPhase = ''
-    py.test --ignore=test/aio/test_shutdown_event.py \
-            --ignore=test/test_shutdown_event.py
-  '';
+  pythonImportsCheck = [ "i3ipc" ];
 
-  meta = with lib; {
+  meta = {
     description = "Improved Python library to control i3wm and sway";
-    homepage = "https://github.com/acrisci/i3ipc-python";
-    license = licenses.bsd3;
-    maintainers = with maintainers; [ vanzef ];
+    homepage = "https://github.com/altdesktop/i3ipc-python";
+    changelog = "https://github.com/altdesktop/i3ipc-python/releases/tag/${src.tag}";
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ vanzef ];
   };
 }

From 4f6020baa6f23fc4fb411dec66b10e71e5db7f4f Mon Sep 17 00:00:00 2001
From: Tom Hunze <dev@thunze.de>
Date: Sat, 30 Aug 2025 23:49:19 +0200
Subject: [PATCH 4/4] python313Packages.i3ipc: improve test stability

---
 pkgs/development/python-modules/i3ipc/default.nix | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/pkgs/development/python-modules/i3ipc/default.nix b/pkgs/development/python-modules/i3ipc/default.nix
index c3458c3bc98fe729d14d8be7cb06f15e5eb97c4e..599966b5e5bdb8b97420744e6339cfdb303f54ef 100644
--- a/pkgs/development/python-modules/i3ipc/default.nix
+++ b/pkgs/development/python-modules/i3ipc/default.nix
@@ -7,6 +7,7 @@
   xlib,
   fontconfig,
   pytestCheckHook,
+  writableTmpDirAsHomeHook,
   pytest-asyncio,
   pytest-timeout,
   pytest-xvfb,
@@ -46,6 +47,7 @@ buildPythonPackage rec {
 
   nativeCheckInputs = [
     pytestCheckHook
+    writableTmpDirAsHomeHook
     pytest-asyncio
     pytest-timeout
     pytest-xvfb
@@ -58,6 +60,9 @@ buildPythonPackage rec {
     # Timeout
     "test/test_shutdown_event.py::TestShutdownEvent::test_shutdown_event_reconnect"
     "test/aio/test_shutdown_event.py::TestShutdownEvent::test_shutdown_event_reconnect"
+    # Flaky
+    "test/test_window.py::TestWindow::test_detailed_window_event"
+    "test/aio/test_workspace.py::TestWorkspace::test_workspace"
   ];
 
   pythonImportsCheck = [ "i3ipc" ];
